src/libasr/pass/replace_symbolic.cpp

```diff
@@ -294,405 +291,196 @@ class ReplaceSymbolicVisitor : public PassUtils::PassVisitor<ReplaceSymbolicVisi
         pass_result.push_back(al, stmt);
     }
 
-    void visit_Assignment(const ASR::Assignment_t &x) {
-        SymbolTable* module_scope = current_scope->parent;
-        if (ASR::is_a<ASR::IntrinsicFunction_t>(*x.m_value)) {
-            ASR::IntrinsicFunction_t* intrinsic_func = ASR::down_cast<ASR::IntrinsicFunction_t>(x.m_value);
-            int64_t intrinsic_id = intrinsic_func->m_intrinsic_id;
-            if (intrinsic_func->m_type->type == ASR::ttypeType::SymbolicExpression) {
-                switch (static_cast<LCompilers::ASRUtils::IntrinsicFunctions>(intrinsic_id)) {
-                    case LCompilers::ASRUtils::IntrinsicFunctions::SymbolicPi: {
-                        std::string new_name = "basic_const_pi";
-                        symbolic_dependencies.push_back(new_name);
-                        if (!module_scope->get_symbol(new_name)) {
-                            std::string header = "symengine/cwrapper.h";
-                            SymbolTable* fn_symtab = al.make_new<SymbolTable>(module_scope);
-
-                            Vec<ASR::expr_t*> args;
-                            args.reserve(al, 1);
-                            ASR::symbol_t* arg = ASR::down_cast<ASR::symbol_t>(ASR::make_Variable_t(
-                                al, x.base.base.loc, fn_symtab, s2c(al, "x"), nullptr, 0, ASR::intentType::In,
-                                nullptr, nullptr, ASR::storage_typeType::Default, ASRUtils::TYPE(ASR::make_CPtr_t(al, x.base.base.loc)),
-                                nullptr, ASR::abiType::BindC, ASR::Public, ASR::presenceType::Required, true));
-                            fn_symtab->add_symbol(s2c(al, "x"), arg);
-                            args.push_back(al, ASRUtils::EXPR(ASR::make_Var_t(al, x.base.base.loc, arg)));
+    ASR::expr_t* handle_argument(Allocator &al, const Location &loc, ASR::expr_t* arg) {
+        if (ASR::is_a<ASR::IntrinsicFunction_t>(*arg) || ASR::is_a<ASR::Cast_t>(*arg)) {
+            if (ASR::is_a<ASR::IntrinsicFunction_t>(*arg)) {
+                this->visit_IntrinsicFunction(*ASR::down_cast<ASR::IntrinsicFunction_t>(arg));
+            } else if (ASR::is_a<ASR::Cast_t>(*arg)) {
+                this->visit_Cast(*ASR::down_cast<ASR::Cast_t>(arg));
+            }
+            ASR::symbol_t* var_sym = current_scope->get_symbol(symengine_stack.pop());
+            return ASRUtils::EXPR(ASR::make_Var_t(al, loc, var_sym));
+        }
+        return arg;
```

Comment:
```suggestion
        if (ASR::is_a<ASR::Var_t>(*arg)) {
            return arg;
        } else if (ASR::is_a<ASR::IntrinsicFunction_t>(*arg)) {
            this->visit_IntrinsicFunction(*ASR::down_cast<ASR::IntrinsicFunction_t>(arg));
        } else if (ASR::is_a<ASR::Cast_t>(*arg)) {
            this->visit_Cast(*ASR::down_cast<ASR::Cast_t>(arg));
        } else {
            LCOMPILERS_ASSERT(false);
        }
        ASR::symbol_t* var_sym = current_scope->get_symbol(symengine_stack.pop());
        return ASRUtils::EXPR(ASR::make_Var_t(al, loc, var_sym));
```
Otherwise, this PR looks great!